src/envoy/http/authnv2/filter.cc

```diff
@@ -0,0 +1,236 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <string>
+
+#include "common/http/utility.h"
+#include "extensions/filters/http/well_known_names.h"
+#include "include/istio/utils/attribute_names.h"
+#include "src/envoy/http/authnv2/filter.h"
+#include "src/envoy/utils/authn.h"
+#include "src/envoy/utils/filter_names.h"
+#include "src/envoy/utils/utils.h"
+
+#include "absl/strings/match.h"
+#include "absl/strings/str_split.h"
+#include "common/json/json_loader.h"
+#include "google/protobuf/struct.pb.h"
+#include "include/istio/utils/attribute_names.h"
+#include "src/envoy/http/jwt_auth/jwt.h"
+
+namespace Envoy {
+namespace Http {
+namespace Istio {
+namespace AuthN {
+
+AuthnV2Filter::~AuthnV2Filter() {}
+
+void AuthnV2Filter::onDestroy() {
+  ENVOY_LOG(debug, "Called AuthnV2Filter : {}", __func__);
+}
+
+// The JWT audience key name
+static const std::string kJwtAudienceKey = "aud";
+// The JWT issuer key name
+static const std::string kJwtIssuerKey = "iss";
+// The key name for the original claims in an exchanged token
+static const std::string kExchangedTokenOriginalPayload = "original_claims";
+
+// Extract JWT claim as a string list.
+// This function only extracts string and string list claims.
+// A string claim is extracted as a string list of 1 item.
+// A string claim with whitespace is extracted as a string list with each
+// sub-string delimited with the whitespace.
+void ExtractStringList(const std::string& key, const Envoy::Json::Object& obj,
+                       std::vector<std::string>* list) {
+  // First, try as string
+  try {
+    // Try as string, will throw execption if object type is not string.
+    const std::vector<std::string> keys =
+        absl::StrSplit(obj.getString(key), ' ', absl::SkipEmpty());
+    for (auto key : keys) {
+      list->push_back(key);
+    }
+  } catch (Json::Exception& e) {
+    // Not convertable to string
+  }
+  // Next, try as string array
+  try {
+    std::vector<std::string> vector = obj.getStringArray(key);
+    for (const std::string v : vector) {
+      list->push_back(v);
+    }
+  } catch (Json::Exception& e) {
+    // Not convertable to string array
+  }
+}
+
+// Helper function to set a key/value pair into Struct.
+static void setKeyValue(::google::protobuf::Struct& data,
+                        const std::string& key, const std::string& value) {
+  (*data.mutable_fields())[key].set_string_value(value);
```

Comment:
TIL stands for "today I learned"? Or something else? Anything need to address here?

Reply:
Yes it is. Thank you!
My major concern is in the main comment: It seems the message has limited field names. Why do you prefer dynamic keys instead of protobuf message with schema?